File: algo/lc.20.cc
class Solution {
public:
    bool isValid(string s) {
        string left="([{";
        string right=")]}";
        stack<char> st;
        
        for (auto c : s) {
            if (left.find(c) != string::npos) { // left parenthis
                st.push(c);
            } else { // right parenthis
                if (st.empty())
                    return false;
                else if (st.top() != left[right.find(c)])
                    return false;
                else
                    st.pop();
            }
        }
        return st.empty();
    }
};

File: algo/lc.20.py
class Solution:
    def isValid(self, s: str) -> bool:
        special = '(){}[]'
        stack = []
        for char in s:
            if char in special:
                if char in '([{':
                    # begin
                    stack.append(char)
                else:
                    # end
                    if not stack:
                        return False
                    if char == ')' and stack[-1] == '(':
                        stack.pop(-1)
                    elif char == ']' and stack[-1] == '[':
                        stack.pop(-1)
                    elif char == '}' and stack[-1] == '{':
                        stack.pop(-1)
                    else:
                        return False
        return not stack
